src/routers/alpha-router/alpha-router.ts

```diff
@@ -877,34 +866,23 @@ export class AlphaRouter
       candidatePools: CandidatePoolsBySelectionCriteria;
     }>[] = [];
 
-    const protocolsSet = new Set(protocols ?? []);
-
-    const [v3gasModel, mixedRouteGasModel] = await Promise.all([
-      this.v3GasModelFactory.buildGasModel({
-        chainId: this.chainId,
-        gasPriceWei,
-        v3poolProvider: this.v3PoolProvider,
-        amountToken: amount.currency.wrapped,
-        quoteToken,
-        v2poolProvider: this.v2PoolProvider,
-        l2GasDataProvider: this.l2GasDataProvider,
-      }),
-      this.mixedRouteGasModelFactory.buildGasModel({
-        chainId: this.chainId,
-        gasPriceWei,
-        v3poolProvider: this.v3PoolProvider,
-        amountToken: amount.currency.wrapped,
-        quoteToken,
-        v2poolProvider: this.v2PoolProvider,
-      }),
-    ]);
+    const protocolsSet = new Set(protocols);
 
-    if (
-      (protocolsSet.size == 0 ||
-        (protocolsSet.has(Protocol.V2) && protocolsSet.has(Protocol.V3))) &&
-      V2_SUPPORTED.includes(this.chainId)
-    ) {
-      log.info({ protocols, tradeType }, 'Routing across all protocols');
+    const [v3gasModel, mixedRouteGasModel] = await this.getGasModels(gasPriceWei, amount.currency.wrapped, quoteToken, metric);
+
+    const noProtocolsSpecified = protocolsSet.size == 0
+    const v3ProtocolSpecified = protocolsSet.has(Protocol.V3)
+    const v2ProtocolSpecified = protocolsSet.has(Protocol.V2)
+    const mixedProtocolSpecified = protocolsSet.has(Protocol.MIXED)
+    const v2SupportedInChain = V2_SUPPORTED.includes(this.chainId)
+    const tradeTypeIsExactInput = tradeType === TradeType.EXACT_INPUT
+    const isMainnet = this.chainId === ChainId.MAINNET
+    const isGorli = this.chainId === ChainId.GÖRLI
+
+    // Maybe Quote V3 - if V3 is specified, or no protocol is specified
+    if (v3ProtocolSpecified || noProtocolsSpecified) {
+      // Open Question: Should we quote V3 if only V2 is specified but v2 is not supported?
```

Comment:
IMHO, if the request wants a v2 quote and v2 is not supported we should return a 4xx? so the client knows that v2 is not supported for this protocol

Reply:
Good call, I think that indirectly happens here: https://github.com/Uniswap/smart-order-router/pull/199/files/040d5685188d9e1d93cc976f72a792947467c4dd#diff-2fe4a092f36b8091262f7035638dc00530816382f3a1cca86927a6fd712e9c04L1033

The boolean logic should in theory be the same as it was before, but easier to understand and simplified

Reply:
Yes we should return no route found in this case, however routing-api should still return 200. I don't think we ever return 400 unless malformed input?